### PR TITLE
[docs] Add a guide on how to support other platforms in Expo modules

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

As of SDK 50, it's possible to write Expo modules for macOS and tvOS. There should be a guide on how to set things up.

# How

Added a new four-step guide

# Test Plan

Use the preview posted below by @expo-bot

### CODE REVIEW

File: docs/pages/modules/supporting-platforms.mdx [hunk 1]
```diff
@@ -0,0 +1,89 @@
+---
+title: Supporting more platforms
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:2]:
```suggestion
title: Additional platform support
```

Let's rephrase the title to use short-form verbs for page titles.

File: docs/pages/modules/supporting-platforms.mdx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+description: Learn how to add support for macOS and tvOS platforms.
+---
+
+import { Step } from '~/ui/components/Step';
+
+Expo Modules API provides the first-class support for Android and iOS, however since all Apple platforms are based on the same foundation and use the same programming language,
+it is possible to target other Apple platforms in the Expo module. At the moment only **macOS** and **tvOS** platforms (also called [Out-of-Tree platforms](https://reactnative.dev/docs/out-of-tree-platforms)) are supported.
+This guide will walk you through adding support for these platforms.
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:10]:
```suggestion
Expo Modules API provides first-class support for Android and iOS. 

At the moment, only **macOS** and **tvOS** platforms (also called [Out-of-Tree platforms](https://reactnative.dev/docs/out-of-tree-platforms)) are supported. However, since all Apple platforms are based on the same foundation and use the same programming language, targeting other Apple platforms in the Expo module is possible. This guide will walk you through adding support for these platforms.
```

- Let Prettier handle the formatting. 
- Also, let's split paragraph where it is described about what platforms are supported

File: docs/pages/modules/supporting-platforms.mdx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+<Step label="1">
+
+### Use the `"apple"` platform in `expo-module.config.json`
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:14]:
```suggestion
## Use the `"apple"` platform in `expo-module.config.json`
```

Let's use an H2 heading here for the correct heading sequence as per our [writing styleguide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#headings).

File: docs/pages/modules/supporting-platforms.mdx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+In order to provide more seamless support for other Apple platforms, Expo SDK 50 introduced a universal `"apple"` platform to instruct the autolinking that the module may support any of the Apple platform
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:16]:
```suggestion
To provide seamless support for other Apple platforms, Expo SDK 50 introduced a universal `"apple"` platform to instruct the autolinking that the module may support any of the Apple platform
```

Let's avoid being too wordy when we can use a single word for the same meaning. Vale also warns and suggests about this:

CleanShot 2024-02-01 at 15 57 47@2x

File: docs/pages/modules/supporting-platforms.mdx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+and the question whether to link the module in the specific CocoaPods target is moved off to the podspec.
+If you have used `"ios"` before, you can safely replace it.
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:18]:
```suggestion
If you have used `"ios"` before, you can safely replace it:
```

File: docs/pages/modules/supporting-platforms.mdx [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
++   "apple": {
++     "modules": ["MyModule"]
++   }
+}
+```
+
+</Step>
+<Step label="2">
+
+### Update the podspec to declare support for other platforms
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:36]:
```suggestion
## Update the podspec to declare support for other platforms
```

File: docs/pages/modules/supporting-platforms.mdx [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+The module's podspec needs to be updated with the supported platforms, otherwise CocoaPods would fail to install the pod on targets for the other platforms.
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:38]:
```suggestion
The module's podspec needs to be updated with a list of the supported platforms. Otherwise, CocoaPods would fail to install the pod on targets for the other platforms.
```

Let's break the sentence for better readability

File: docs/pages/modules/supporting-platforms.mdx [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+```diff Module's podspec
+- s.platform       = :ios, '13.4'
++ s.platforms = {
++   :ios => '13.4',
++   :tvos => '13.4',
++   :osx => '10.15'
++ }
+```
+
+Note that any changes in the podspec require running `pod install` to have an effect.
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:50]:
```suggestion
Any changes in the podspec require running `pod install` to have an effect.
```

File: docs/pages/modules/supporting-platforms.mdx [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+</Step>
+
+<Step label="3">
+
+### Set up `react-native-macos` or `react-native-tvos` in the app
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:56]:
```suggestion
## Set up `react-native-macos` or `react-native-tvos` in the app
```

File: docs/pages/modules/supporting-platforms.mdx [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+If you are writing a local module and your app is already set up, you can skip this step.
+Otherwise, you will need to set up your app or the example app if you are writing a standalone (non-local) module.
+
+- **For macOS**: follow the official [Install React Native for macOS](https://microsoft.github.io/react-native-windows/docs/rnm-getting-started#install-the-macos-extension) guide from `react-native-macos` documentation.
+- **For tvOS**: follow the instructions in the [react-native-tvos](https://github.com/react-native-tvos/react-native-tvos) repository.
+
+</Step>
+<Step label="4">
+
+### Review the code for using APIs not supported on these platforms
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:67]:
```suggestion
## Review the code for using APIs not supported on these platforms
```

File: docs/pages/modules/supporting-platforms.mdx [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+Platform APIs may differ between Apple platforms. The most noticeable difference comes from relying on different UI frameworks – `UIKit` on iOS/tvOS and `AppKit` on macOS.
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:69]:
```suggestion
Platform APIs may differ between Apple platforms. The most noticeable difference comes from relying on different UI frameworks &mdash;`UIKit` on iOS/tvOS and `AppKit` on macOS.
```

File: docs/pages/modules/supporting-platforms.mdx [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+Both `react-native-macos` and `expo-modules-core` provide some aliases and polyfills so that you can reference to `UIKit` classes on macOS target (for example `UIView` is an alias to `NSView`, `UIApplication` is an alias to `NSApplication`),
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:71]:
```suggestion
Both `react-native-macos` and `expo-modules-core` provide aliases and polyfills to reference`UIKit` classes on macOS target (for example, `UIView` is an alias to `NSView`, `UIApplication` is an alias to `NSApplication`),
```

We can a bit concise here.

File: docs/pages/modules/supporting-platforms.mdx [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+but it's usually not enough for iOS-first libraries to support other platforms out of the box. You may need to write conditionally compiled code that uses different implementations depending on the platform.
+
+In order to do this, you can use Swift compiler directives with the `os` condition, which includes a given piece of code when our app is being built for a specific platform.
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:74]:
```suggestion
To do this, use Swift compiler directives with the `os` condition, which includes a given piece of code when our app is being built for a specific platform.
```

File: docs/pages/modules/supporting-platforms.mdx [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+That, combined with the `#if` and `#else` directives, essentially lets you set up platform-specific branches within the cross-platform code.
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:75]:
```suggestion
In combination with the `#if` and `#else` directives, lets you set up platform-specific branches within the cross-platform code.
```

File: docs/pages/modules/supporting-platforms.mdx [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+#elseif os(macOS)
+  // macOS implementation
+#elseif os(tvOS)
+  // tvOS implementation
+#endif
+```
+
+</Step>
+
+Congratulations! Your module is now ready to be used on Out-of-Tree platform.
```
Comment (amandeepmittal) [inline docs/pages/modules/supporting-platforms.mdx:89]:
```suggestion
Your module is now ready to be used on Out-of-Tree platform.
```

**Review by lukmccall — APPROVED** [review]

File: docs/pages/modules/supporting-platforms.mdx (hunk 2 above)
Comment (tsapeta) [reply docs/pages/modules/supporting-platforms.mdx:10]: I think your suggestion has a little bit different meaning and is confusing for me. What about this?

```suggestion
Expo Modules API provides first-class support for Android and iOS. However, since all Apple platforms are based on the same foundation and use the same programming language, targeting other [Out-of-Tree platforms](https://reactnative.dev/docs/out-of-tree-platforms) in the Expo module is possible. At the moment, only **macOS** and **tvOS** platforms are supported. This guide will walk you through adding support for these platforms.
```

File: docs/pages/modules/supporting-platforms.mdx (hunk 15 above)
Comment (tsapeta) [reply docs/pages/modules/supporting-platforms.mdx:89]: I'll remove it from here, but as far as I remember we have this in some other guides too

File: docs/pages/modules/supporting-platforms.mdx (hunk 2 above)
Comment (amandeepmittal) [reply docs/pages/modules/supporting-platforms.mdx]: I see your point. Yes, this sounds good to me 👍

File: docs/pages/modules/additional-platform-support.mdx [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+<Step label="3">
+
+## Set up `react-native-macos` or `react-native-tvos` in the app
+
+If you are writing a local module and your app is already set up, you can skip this step.
+Otherwise, you will need to set up your app or the example app if you are writing a standalone (non-local) module.
+
+- **For macOS**: follow the official [Install React Native for macOS](https://microsoft.github.io/react-native-windows/docs/rnm-getting-started#install-the-macos-extension) guide from `react-native-macos` documentation.
+- **For tvOS**: follow the instructions in the [react-native-tvos](https://github.com/react-native-tvos/react-native-tvos) repository.
```
Comment (douglowder) [inline docs/pages/modules/additional-platform-support.mdx:62]:
```suggestion
- **For tvOS**: follow the instructions in the [react-native-tvos](https://github.com/react-native-tvos/react-native-tvos) repository. If you are building an Expo app, you should also follow the instructions in our [Building Expo apps for TV guide](https://docs.expo.dev/guides/building-for-tv/).
```

**Review by douglowder — APPROVED** [review]
One suggestion for TV

File: docs/pages/modules/additional-platform-support.mdx (hunk 16 above)
Comment (Simek) [reply docs/pages/modules/additional-platform-support.mdx:62]:
```suggestion
- **For tvOS**: follow the instructions in the [react-native-tvos](https://github.com/react-native-tvos/react-native-tvos) repository. If you are building an Expo app, you should also follow the instructions in our [Building Expo apps for TV guide](/guides/building-for-tv/).
```

Version of suggestion with an internal link to other doc, to avoid full re-render.

File: docs/pages/modules/additional-platform-support.mdx [hunk 17]
```diff
@@ -0,0 +1,89 @@
+---
+title: Additional platform support
+description: Learn how to add support for macOS and tvOS platforms.
+---
+
+import { Step } from '~/ui/components/Step';
+
+Expo Modules API provides the first-class support for Android and iOS, however since all Apple platforms are based on the same foundation and use the same programming language,
+it is possible to target other Apple platforms in the Expo module. At the moment only **macOS** and **tvOS** platforms (also called [Out-of-Tree platforms](https://reactnative.dev/docs/out-of-tree-platforms)) are supported.
+This guide will walk you through adding support for these platforms.
```
Comment (tsapeta) [inline docs/pages/modules/additional-platform-support.mdx:10]:
```suggestion
Expo Modules API provides first-class support for Android and iOS. However, since all Apple platforms are based on the same foundation and use the same programming language, targeting other [Out-of-Tree platforms](https://reactnative.dev/docs/out-of-tree-platforms) in the Expo module is possible. At the moment, only **macOS** and **tvOS** platforms are supported.
This guide will walk you through adding support for these platforms.
```

**Review by amandeepmittal — APPROVED** [review]
Thank you for adding this guide!
